# Week 5 (2/28-3/6)

## [Pandas](../Tools/Pandas/pandas_toc.rst)

* GroupBy
* Index operations

## Resorces

### 1. Seaborn sample data 

## Exercises

All exercises below use data on restaurant visits. A DataFrame with this data can be created as follows:

In [23]:
import seaborn as sns
tips = sns.load_dataset("tips")
tips.head(5)

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

### Exercise 1

How many people visited the restaurant each day of the week? 

**Check:** There were 152 visitors on Thursday.

In [24]:
# "nbsphinx": "hidden"

tips.groupby(by="day")["size"].sum()

day
Thur    152
Fri      40
Sat     219
Sun     216
Name: size, dtype: int64

### Exercise 2

What was the average bill paid each day of the week? 

**Check:** <span class="tex2jax_ignore">The average bill on Thursday was $17.48</span>.

In [25]:
# "nbsphinx": "hidden"

tips.groupby(by="day")["total_bill"].mean()

day
Thur    17.682742
Fri     17.151579
Sat     20.441379
Sun     21.410000
Name: total_bill, dtype: float64

### Exercise 3

For each restaurant visit, compute  what percentage of the total bill was the tip, and then compute the average value of this percentage for male and for female customers. 

**Check:** The average tip for males was 15.76%.

In [26]:
# "nbsphinx": "hidden"

tips["tip_perc"] = 100*tips["tip"]/tips["total_bill"]
tips.groupby(by="sex")["tip_perc"].mean()

sex
Male      15.765055
Female    16.649074
Name: tip_perc, dtype: float64

### Exercise 4

Compute a DateFrame with 10 rows selected from the tips data: 5 rows with the the 5 highest total bills for smokers and 5 rows with the 5 highest total bills for non-smokers


**Check:** The 5 highest bills for smokers were <span class="tex2jax_ignore">$50.81, $43.35, $44.30, $43.11 and $40.55.

In [27]:
# "nbsphinx": "hidden"

def large_bills(df):

    df = df.sort_values(by="total_bill", ascending=False)
    return df.head(5)

big_bills = tips.groupby("smoker").apply(large_bills)
big_bills

total_bill    tip     sex smoker   day    time  size   tip_perc
smoker                                                                     
Yes    170       50.81  10.00    Male    Yes   Sat  Dinner     3  19.681165
       182       45.35   3.50    Male    Yes   Sun  Dinner     3   7.717751
       102       44.30   2.50  Female    Yes   Sat  Dinner     3   5.643341
       197       43.11   5.00  Female    Yes  Thur   Lunch     4  11.598237
       184       40.55   3.00    Male    Yes   Sun  Dinner     2   7.398274
No     212       48.33   9.00    Male     No   Sat  Dinner     4  18.621974
       59        48.27   6.73    Male     No   Sat  Dinner     4  13.942407
       156       48.17   5.00    Male     No   Sun  Dinner     6  10.379905
       142       41.19   5.00    Male     No  Thur   Lunch     5  12.138869
       23        39.42   7.58    Male     No   Sat  Dinner     4  19.228818

### Exercise 5

Create a DataFrame with 4 rows selected from the tips data: a row with the highest total bill for a male smoker, a row with the highest total bill for a male non-smoker, and the same for a female smokers and a female non-smokers. 


**Check:** <span class="tex2jax_ignore">The highest bills were $50.81 for male smoker, $48.33 for a male non-smoker, $44.30 for a female smoker and $35.83 for a female non-smoker.</span>

In [31]:
# "nbsphinx": "hidden"

def max_bills(df):

    return df.loc[df["total_bill"].idxmax()]

big_bills = tips.groupby(["sex", "smoker"]).apply(max_bills)
big_bills

total_bill    tip     sex smoker  day    time  size   tip_perc
sex    smoker                                                                
Male   Yes          50.81  10.00    Male    Yes  Sat  Dinner     3  19.681165
       No           48.33   9.00    Male     No  Sat  Dinner     4  18.621974
Female Yes          44.30   2.50  Female    Yes  Sat  Dinner     3   5.643341
       No           35.83   4.67  Female     No  Sat  Dinner     3  13.033771

### Exercise 6

Create a DataFrame with two rows indicating if a customer was a male or a female, and columns indicating which day of the week the customer visited the restaurant. The value in a given row and column should be the amount of the largest tip that a customer of a given sex and visiting on a given day paid. 

**Check:** <span class="tex2jax_ignore">The highest tip paid on Thursday was $6.70 for males and $5.17 for females.</span>

In [22]:
# "nbsphinx": "hidden"

df.groupby(["sex", "day"])["tip"].max().unstack()

day     Thur   Fri   Sat  Sun
sex                          
Male    6.70  4.73  10.0  6.5
Female  5.17  4.30   6.5  5.2